solana/write-account/src/lib.rs

```diff
@@ -9,145 +12,227 @@ solana_program::entrypoint!(process_instruction);
-    let (wr, rd) = match accounts {
-        [wr, rd, ..] if wr.is_writable => Ok((wr, rd)),
-        [_, _, ..] => Err(ProgramError::InvalidAccountData),
-        _ => Err(ProgramError::NotEnoughAccountKeys),
-    }?;
-
-    let algo = data.unshift().map_or(0, |n| *n);
-    let offset = data.unshift_n().map_or(Ok(0), usize_from_bytes)?;
-    let start = data.unshift_n().map_or(Ok(0), usize_from_bytes)?;
-    let end = data.unshift_n().map(usize_from_bytes).transpose()?;
-    if !data.is_empty() {
-        return Err(ProgramError::InvalidAccountData);
+/// Verifies and sets up the write account.
+///
+/// Firstly, checks that the write accounts address corresponds to the PDA which
+/// we’d get by using `[payer.key, seed]` as seeds with given `bump`.  If it
+/// doesn’t, returns `InvalidSeeds` error.
+///
+/// Secondly, if the account doesn’t exist, creates it with size of `size`.
+/// Note that due to Solana limitations, `size` may be at most 10 KiB in this
+/// case (see [`solana_program::entrypoint::MAX_PERMITTED_DATA_INCREASE`]).
+///
+/// Otherwise, checks if account’s size it at least `size`.  If it isn’t,
+/// resizes the account (see [`AccountInfo::realloc`]).  Again, due to Solana’s
+/// limitations, account may grow by at most 10 KiB.  To remain rent exempt,
+/// this may lead to lamports being transferred from `payer` to the
+/// `write_account`.
+///
+fn setup_write_account<'info>(
+    program_id: &Pubkey,
+    payer: &AccountInfo<'info>,
+    write_account: &AccountInfo<'info>,
+    seed: &[u8],
+    bump: u8,
+    size: usize,
+    system: Option<&AccountInfo<'info>>,
+) -> Result {
+    let seeds = &[payer.key.as_ref(), seed, &[bump]];
+    match Pubkey::create_program_address(seeds, program_id) {
+        Ok(pda) if &pda == write_account.key => (),
+        _ => return Err(ProgramError::InvalidSeeds),
     }
 
-    let mut dst = wr.try_borrow_mut_data()?;
-    let dst = dst.get_mut(offset..).ok_or(ProgramError::AccountDataTooSmall)?;
-    let end = end.map_or_else(|| rd.try_data_len(), Ok)?;
-    let src = rd.try_borrow_data()?;
-    let src = src.get(start..end).ok_or(ProgramError::AccountDataTooSmall)?;
-
-    match algo {
-        0 => handle_copy_null(dst, src),
-        _ => Err(ProgramError::InvalidInstructionData),
+    let lamports = write_account.lamports();
+    let get_required_lamports =
+        || Rent::get().map(|rent| rent.minimum_balance(size));
+
+    if lamports == 0 {
+        // If the account has zero lamports it needs to be created first.
+        let _ = system.ok_or(ProgramError::NotEnoughAccountKeys)?;
+        let instruction = solana_program::system_instruction::create_account(
+            payer.key,
+            write_account.key,
+            get_required_lamports()?,
+            size as u64,
+            program_id,
+        );
+        solana_program::program::invoke_signed(
+            &instruction,
+            &[payer.clone(), write_account.clone()],
```

Comment:
I feel creating account requires system program.
```suggestion
            &[payer.clone(), write_account.clone(), system_program.clone()],
```

Reply:
Solana is weird. The system program account must be there passed to the smart contract invocation, but it’s not needed to in the invocation. See [example in documentation](https://docs.rs/solana-program/latest/solana_program/system_instruction/fn.create_account.html#example-on-chain-program). I’m also double-checking in #215.

Reply:
This is strange.

Reply:
Anchor Test in #215 passed so this PR is correct. System program doesn’t need to be passed in the invocation.